daemon/containerd/resolver.go

```diff
@@ -1,35 +1,89 @@
 package containerd
 
 import (
+	"net/http"
+	"strings"
+
 	"github.com/containerd/containerd/remotes"
 	"github.com/containerd/containerd/remotes/docker"
 	registrytypes "github.com/docker/docker/api/types/registry"
 	"github.com/docker/docker/registry"
 	"github.com/sirupsen/logrus"
 )
 
-func newResolverFromAuthConfig(authConfig *registrytypes.AuthConfig) remotes.Resolver {
-	opts := []docker.RegistryOpt{}
-	if authConfig != nil {
-		cfgHost := registry.ConvertToHostname(authConfig.ServerAddress)
-		if cfgHost == registry.IndexHostname {
-			cfgHost = registry.DefaultRegistryHost
+func (i *ImageService) newResolverFromAuthConfig(authConfig *registrytypes.AuthConfig) (remotes.Resolver, docker.StatusTracker) {
+	tracker := docker.NewInMemoryTracker()
+	hostsFn := i.registryHosts.RegistryHosts()
+
+	hosts := hostsWrapper(hostsFn, authConfig, i.registryService)
+
+	return docker.NewResolver(docker.ResolverOptions{
+		Hosts:   hosts,
+		Tracker: tracker,
+	}), tracker
+}
+
+func hostsWrapper(hostsFn docker.RegistryHosts, authConfig *registrytypes.AuthConfig, regService registry.Service) docker.RegistryHosts {
+	return docker.RegistryHosts(func(n string) ([]docker.RegistryHost, error) {
+		hosts, err := hostsFn(n)
+		if err != nil {
+			return nil, err
 		}
-		authorizer := docker.NewDockerAuthorizer(docker.WithAuthCreds(func(host string) (string, string, error) {
-			if cfgHost != host {
-				logrus.WithField("host", host).WithField("cfgHost", cfgHost).Warn("Host doesn't match")
-				return "", "", nil
-			}
-			if authConfig.IdentityToken != "" {
-				return "", authConfig.IdentityToken, nil
+
+		for idx, host := range hosts {
+			if host.Authorizer == nil {
+				var opts []docker.AuthorizerOpt
+				if authConfig != nil {
+					opts = append(opts, authorizationCredsFromAuthConfig(*authConfig))
+				}
+				host.Authorizer = docker.NewDockerAuthorizer(opts...)
+
+				isInsecure := regService.IsInsecureRegistry(host.Host)
+				if host.Client.Transport != nil && isInsecure {
+					host.Client.Transport = httpFallback{super: host.Client.Transport}
+				}
+				hosts[idx] = host
 			}
```

Comment:
Mutate the slice indices in-place instead of copying twice?
```suggestion
		for i := range hosts {
			if hosts[i].Authorizer == nil {
				var opts []docker.AuthorizerOpt
				if authConfig != nil {
					opts = append(opts, authorizationCredsFromAuthConfig(*authConfig))
				}
				hosts[i].Authorizer = docker.NewDockerAuthorizer(opts...)

				isInsecure := regService.IsInsecureRegistry(host.Host)
				if host.Client.Transport != nil && isInsecure {
					hosts[i].Client.Transport = httpFallback{super: hosts[i].Client.Transport}
				}
			}
```